### PR TITLE
Fix #1321 and #1459 and #171 and #1320 and #1456

### LINKED ISSUE #1320: Moved to other screen maximized window un-maximizes on first screen
Output of `awesome --version`:

    awesome v4.0 (Harder, Better, Faster, Stronger)
     • Compiled against Lua 5.2.4 (running with Lua 5.2)
     • D-Bus support: ✔
     • execinfo support: ✔
     • RandR 1.5 support: ✔
     • LGI version: 0.9.1

**How to reproduce the issue:**
1. Have two screens. Layout doesn't matter.
2. `S-Ret S-m S-o` on first screen to:
    * create terminal window (actually any window will do)
    * maximize it, and
    * send to other screen.
3. `S-m` on other screen when focused on terminal window from previous step.

**Actual result:**
Terminal window un-maximizes *on the first screen*.
Focus remains on the second screen.

**Expected result:**
I don't really know, but I personally expect window to un-maximize on the screen it currently belongs to.

### LINKED ISSUE #1321: Float mode un-maximization doesn't restore window width
Output of `awesome --version`:
```
awesome v4.0 (Harder, Better, Faster, Stronger)
 • Compiled against Lua 5.2.4 (running with Lua 5.2)
 • D-Bus support: ✔
 • execinfo support: ✔
 • RandR 1.5 support: ✔
 • LGI version: 0.9.1
```
**How to reproduce the issue:**
1. Have tag in floating layout.
2. `S-Ret S-m S-m` to create terminal (e.g., any window will do), maximize and un-maximize it.

**Actual result:**
Window's height is restored to what it was before maximization, but width is not.

**Expected result:**
Width is also returned to what it was before.

### PR DESCRIPTION
**Pull request by Elv13**
Before: to everything twice including the request::geometry
Now: do it once

Edit: also create a real `maximized` property, spliting the state across 2 properties is a source of corner cases.

@taptap @cagelight can you test?

Fix #1320 #1321 #171 #1459

### CODE REVIEW

File: objects/client.h [hunk 1]
```diff
@@ -47,6 +47,12 @@ typedef enum {
     CLIENT_TITLEBAR_COUNT = 4
 } client_titlebar_t;
 
+typedef enum {
+    CLIENT_MAXIMIZED_NONE = 0 << 0,
+    CLIENT_MAXIMIZED_V    = 1 << 0,
+    CLIENT_MAXIMIZED_H    = 1 << 1,
+} client_maximized_t;
```
Comment (psychon) [inline objects/client.h]: Not needed in the header. Can't this be right in the C file?

Comment (Elv13) [inline objects/client.h]: Done

**Comment by Elv13** [comment]
Ah, as I think of it, its a behavior change for the 0.1 users who set both manually. I should change this to be additive instead of overwriting.

File: objects/client.c [hunk 2]
```diff
@@ -1919,38 +1920,68 @@ client_get_maximized(client_t *c)
...
+void
+client_set_maximized_common(lua_State *L, int cidx, bool s, const char* type, const int val)
+{
+    client_t *c = luaA_checkudata(L, cidx, &client_class);
+
+    /* Store the current and next state on 2 bit */
+    const client_maximized_t current = (
+        (c->maximized_vertical   ? CLIENT_MAXIMIZED_V : CLIENT_MAXIMIZED_NONE)|
+        (c->maximized_horizontal ? CLIENT_MAXIMIZED_H : CLIENT_MAXIMIZED_NONE)
+    );
```
Comment (psychon) [inline objects/client.c]: Well, ok. One thing: Why two bits instead of just using two different variables for this? GCC can likely figure out nice and fast code anyway (not that it matters much here...) and I think the result would be more readable.

**Comment by taptap** [comment]
It still restores on the monitor where it was maximized.
However, it's not widened to whole width anymore.

File: objects/client.c (hunk 2 above)
Comment (Elv13) [inline objects/client.c]: The property macro like booleans

**Comment by Bew78LesellB** [comment]
It works well for me, this fix #1459, thanks !

File: objects/client.c [hunk 3]
```diff
@@ -1919,38 +1926,68 @@ client_get_maximized(client_t *c)
...
+    const client_maximized_t next = s ? val : CLIENT_MAXIMIZED_NONE;
```
Comment (psychon) [inline objects/client.c]: Likely makes no difference, but this could be moved into the caller. Instead of having a boolean and a mask, this function could directly get the `next` value.
</bikeshedding>

File: tests/test-maximize.lua [hunk 4]
```diff
@@ -99,6 +100,47 @@ local steps = {
...
+        c.maximized = false
+
+        return true
+    end,
+    function()
+        local c = client.get()[1]
+
+        if c.maximized then
+            return
+        end
```
Comment (psychon) [inline tests/test-maximize.lua]: Is this `if` necessary? Setting `c.maximized` should be instantaneous (so this could even be done in the previous step)

File: ewmh.c [hunk 5]
```diff
@@ -216,6 +216,7 @@ ewmh_init(void)
     luaA_class_connect_signal(L, &client_class, "property::fullscreen" , ewmh_client_update_hints);
     luaA_class_connect_signal(L, &client_class, "property::maximized_horizontal" , ewmh_client_update_hints);
     luaA_class_connect_signal(L, &client_class, "property::maximized_vertical" , ewmh_client_update_hints);
+    luaA_class_connect_signal(L, &client_class, "property::maximized" , ewmh_client_update_hints);
```
Comment (psychon) [inline ewmh.c]: So... a client can be maximized vertically, maximized horizontally and at the same time not be maximized? Put differently, these are now three completely independent booleans? And all code that wants to check for `maximized_vertical` should now check `maximized_vertical or maximized`, because the C code no longer does so for you? That seems rather odd and I don't like it at all.

How about deprecating the horizontal/vertical maximization instead and only have a single "maximized" state instead?

File: lib/awful/placement.lua [hunk 6]
```diff
@@ -1422,6 +1422,11 @@ function placement.restore(d, args)
 
     if not memento then return false end
 
+    -- Apparently, 0x0 did happen to at least 1 person. Officially can't
+    -- happen, but can't be too careful.
+    if memento.width  and memento.width < 1 then return end
+    if memento.height and memento.height < 1 then return end
```
Comment (psychon) [inline lib/awful/placement.lua]: How about moving this into the creation of a memento instead, so that such an invalid memento cannot be created at all?

However, I am curious how this could happen. This seems to require `c:geometry()` to return 0 for width/height. This can only happen while the client is still being managed, because there is a time frame while the value is still 0 before the values coming from the X11 server are applied. This can only happen during the following signals: `new`, `property::x`, `property::y`, `property::width`, `property::screen`, `property::window`.

An (in my opinion) even better proposal: Instead of working around this in Lua, change the C code so that it emits the signals later (except for `new`, but no one should use that one anyway)?
Basically:
```diff
diff --git a/objects/client.c b/objects/client.c
index ec4ab424b..9f6bd138b 100644
--- a/objects/client.c
+++ b/objects/client.c
@@ -1405,8 +1405,6 @@ client_manage(xcb_window_t w, xcb_get_geometry_reply_t *wgeom, xcb_get_window_at
      * (Else, reparent could cause an UnmapNotify) */
     xcb_change_window_attributes(globalconf.connection, w, XCB_CW_EVENT_MASK, select_input_val);
 
-    luaA_object_emit_signal(L, -1, "property::window", 0);
-
     /* The frame window gets the border, not the real client window */
     xcb_configure_window(globalconf.connection, w,
                          XCB_CONFIG_WINDOW_BORDER_WIDTH,
@@ -1429,15 +1427,16 @@ client_manage(xcb_window_t w, xcb_get_geometry_reply_t *wgeom, xcb_get_window_at
 
     /* Store initial geometry and emits signals so we inform that geometry have
      * been set. */
-#define HANDLE_GEOM(attr) \
-    c->geometry.attr = wgeom->attr; \
-    luaA_object_emit_signal(L, -1, "property::" #attr, 0);
-HANDLE_GEOM(x)
-HANDLE_GEOM(y)
-HANDLE_GEOM(width)
-HANDLE_GEOM(height)
-#undef HANDLE_GEOM
-
+    c->geometry.x = wgeom->x;
+    c->geometry.y = wgeom->y;
+    c->geometry.width = wgeom->width;
+    c->geometry.height = wgeom->height;
+
+    luaA_object_emit_signal(L, -1, "property::x", 0);
+    luaA_object_emit_signal(L, -1, "property::y", 0);
+    luaA_object_emit_signal(L, -1, "property::width", 0);
+    luaA_object_emit_signal(L, -1, "property::height", 0);
+    luaA_object_emit_signal(L, -1, "property::window", 0);
     luaA_object_emit_signal(L, -1, "property::geometry", 0);
 
     /* Set border width */
```

File: lib/awful/ewmh.lua [hunk 7]
```diff
@@ -270,6 +275,10 @@ function ewmh.geometry(c, context, hints)
             props.honor_workarea = honor_default
         end
 
+        if props.honor_padding == nil and props.honor_workarea then
+            props.honor_padding = beautiful.maximized_honor_padding ~= false
+        end
```
Comment (psychon) [inline lib/awful/ewmh.lua]: The commit message says that this is specific to maximisation, but I fail to see that in the code...?

File: lib/awful/ewmh.lua [hunk 8]
```diff
@@ -284,7 +279,6 @@ client.connect_signal("request::tag", ewmh.tag)
 client.connect_signal("request::urgent", ewmh.urgent)
 client.connect_signal("request::geometry", ewmh.geometry)
 client.connect_signal("property::border_width", geometry_change)
-client.connect_signal("property::geometry", geometry_change)
```
Comment (psychon) [inline lib/awful/ewmh.lua]: ...sounds like the function name no longer matches its job (although I am not quite sure what that job is). Could you clean that up?

File: lib/awful/ewmh.lua [hunk 9]
```diff
@@ -42,18 +42,13 @@ local function geometry_change(window)
...
-        window:geometry { width = g.width - 2*bw, x = g.x }
-    end
-    if window.fullscreen then
-        window.border_width = 0
-        window:geometry(window.screen.geometry)
+    for prop in ipairs {
+        "fullscreen", "maximized", "maximized_vertical", "maximized_horizontal"
+    } do
+        if window[prop] then
+            window:emit_signal("request::geometry", prop, {})
```
Comment (psychon) [inline lib/awful/ewmh.lua]: Should the last argument somehow indicate that this is just re-applying the way things should be?

File: tests/test-maximize.lua [hunk 10]
```diff
@@ -82,10 +82,11 @@ local steps = {
...
         assert(c.fullscreen)
-        assert(new_geo.x-bw==sgeo.x)
-        assert(new_geo.y-bw==sgeo.y)
-        assert(new_geo.x+new_geo.width+bw==sgeo.x+sgeo.width)
-        assert(new_geo.y+new_geo.height+bw==sgeo.y+sgeo.height)
+
+        assert(new_geo.x==sgeo.x)
+        assert(new_geo.y==sgeo.y)
+        assert(new_geo.x+new_geo.width+2*bw==sgeo.x+sgeo.width)
+        assert(new_geo.y+new_geo.height+2*bw==sgeo.y+sgeo.height)
```
Comment (psychon) [inline tests/test-maximize.lua]: Do I understand correctly that the code doing ` window:geometry(window.screen.geometry)` was actually wrong (since it ignored the border width) and the wrongness managed to end up in a test? Yuck.

Comment (Elv13) [inline tests/test-maximize.lua]: That code path wasn't of my making ;) (but the test is). Actually, after thinking about it, it was probably done on purpose. Done this way, you get "real" fullscreen. This was initially why I left it that way.  But now it conflicts with the request API, so it has to use the standard code path or these issues will keep piling up.

File: ewmh.c (hunk 5 above)
Comment (Elv13) [inline ewmh.c]: The previous version of the patch kept the 2bit==maximized concept, but it doesn't work after all.

*problem 1:*
When you maximized something, all 3 signals were sent. If some code handled "maximized" a "maximized_horizontally" (or all 3), then it was executed 3 times.

*problem 1.1:*
`request::geometry` was sent twice (before this PR) when you did a single operation. This is why it failed to serialize the size.

*problem 2:*
If you had the 3 maximize button in the titlebar, than where you toggled maximize twice, it would override the other 2 and do->undo would not restore the state as it was.

So, having a split state made most operation *more* completed than having 3 independent states. Also, have 3 booleans ensure that when changing one of them, only 1 signal is emitted. The previous code worked, I am not denying that, but it had the "design quirks build-in" while the request API blueyed and me implemented is designed to unify "concepts" into a single handler. It *could* be fixed in the handler, but I would rather have a simplified pipeline. This is what this patch does. Most users use maximize and not the other 2, so simplifying that one while making the other 2 a little more complex doesn't seem such a bad idea.

> How about deprecating the horizontal/vertical maximization instead and only have a single "maximized" state instead?

I don't think its necessary. Once this patch becomes stable, it wont need to be changed. I probably should merge `fullscreen` into this function as well. I will if I ever attack #758 and [insert z-index limitations issue I can't find right now].

File: tests/test-maximize.lua (hunk 10 above)
Comment (psychon) [inline tests/test-maximize.lua]: ...so, does that mean that the next issue will be the "fullscreen" is no longer fullscreen? That's not good. Fullscreen should hide the border (=set border width to zero).

Comment (Elv13) [inline tests/test-maximize.lua]: Done

**Comment by Elv13** [comment]
I addressed many comments and added the fix for #1320. I don't expect this to be approved without another round of changes, but at least this help the pile of open PRs I have right now.

Again, I want to point out that adding another maximized mode was made to simplify the code, even if it makes the C code is little bit uglier.

File: objects/client.c [hunk 11]
```diff
@@ -1962,21 +1953,28 @@ client_set_maximized_common(lua_State *L, int cidx, bool s, const char* type, co
...
+        (c->maximized_vertical   ? CLIENT_MAXIMIZED_V    : CLIENT_MAXIMIZED_NONE)|
+        (c->maximized_horizontal ? CLIENT_MAXIMIZED_H    : CLIENT_MAXIMIZED_NONE)|
+        (c->maximized            ? CLIENT_MAXIMIZED_BOTH : CLIENT_MAXIMIZED_NONE)
     );
-    const client_maximized_t next = s ? val : CLIENT_MAXIMIZED_NONE;
+    client_maximized_t next = s ? (val | current) : (current & (~val));
+
+    /* When both are already set during startup, assume `maximized` is true*/
+    if (next == (CLIENT_MAXIMIZED_H|CLIENT_MAXIMIZED_V) && !globalconf.loop)
+        next = CLIENT_MAXIMIZED_BOTH;
```
Comment (psychon) [inline objects/client.c]: Shouldn't this by `|=` instead of `=`?
Or should we invent our own `MAXIMIZED` state for EWMH and use that one here? If so, feel free to open a todo-issue assigned to me telling me to get rid of this again.

File: lib/awful/ewmh.lua [hunk 12]
```diff
@@ -280,7 +284,18 @@ function ewmh.geometry(c, context, hints)
...
+            props.ignore_border_width = true
+        end
+
         aplace[context](c, props)
+
+        -- Remove the border to get a "real" fullscreen.
+        if original_context == "fullscreen" and beautiful.fullscreen_hide_border ~= false then
+            repair_geometry_lock = true
+            c.border_width = 0
+            repair_geometry_lock = false
```
Comment (psychon) [inline lib/awful/ewmh.lua]: shouldn't this be:
```
local original = repair_geometry_lock
repair_geometry_lock = true
c.border_width = 0
repair_geometry_lock = original
```
After all, I think that this function here will be called by `repair_geometry()`.

File: lib/awful/rules.lua [hunk 13]
```diff
@@ -363,8 +363,8 @@ function rules.extra_properties.placement(c, value)
     local ty = type(value)
 
     local args = {
-        honor_workarea = true,
-        honor_padding  = true
+        honor_workarea = props.honor_workarea ~= false,
+        honor_padding  = props.honor_padding ~= false
```
Comment (psychon) [inline lib/awful/rules.lua]: Undocumented?

File: lib/awful/placement.lua [hunk 14]
```diff
@@ -1422,11 +1422,33 @@ function placement.restore(d, args)
...
-    memento.screen = nil --TODO use it
+    local x, y = memento.x, memento.y
+
+    -- Some people consider that once moved to another screen, then
+    -- the memento needs to be upgraded. For now this is only true for
+    -- maximization until someone complains.
+    if memento.screen and memento.screen.valid and args.context == "maximize"
+      and d.screen and get_screen(memento.screen) ~= get_screen(d.screen) then
+        -- Use the absolute geometry as the memento also does
+        local oldsgeo = memento.screen.geometry
```
Comment (psychon) [inline lib/awful/placement.lua]: Should the screen's geometry be saved inside the memento? That way this would also work if the original screen has since be removed.
Also, the screen's geometry could have changed, so the `x`,`y` saved in there need to be fixed up anyway (the `get_screen(memento.screen) ~= get_screen(d.screen) ` should be dropped), I think.

**Comment by psychon** [comment]
Tentative :+1: besides the comments above.

File: objects/client.c (hunk 11 above)
Comment (Elv13) [inline objects/client.c]: > Shouldn't this by |= instead of =?

Nah, this is one of the possible "bad side effect" of this. If someone manually set both `maximized_[*]`, then it's interpreted as maximized on on restart. The opposite, if `=|` was used and the user had no shortcut for `maximized_[*]`, then he/she/it would be stuck with a triple maximized client with no way to turn it off. The "issue" itself already exist, but isn't "as bad" since the default maximized shortcut is currently "destructive" in the sense toggling it twice will reset both maximization. From now on, if one of the `maximize_*` is set and not the other, then it will be harder for the user to turn it off. The `=` just makes sure it doesn't happen because of this change.

At first I was thinking your idea of another maximized xprop was a good idea, but thinking about it again, I think a better course of action might be to reset fullscreen and maximization properties when restarting. It is a behavior change compared to previous version and someone, somewhere, will notice. However it simplifies the algorithm and users don't restart Awesome as often as they used to, so it wont be as noticeable as before. Anyway, when you lose your layouts and stack order, keeping maximization cause a confusing result.

File: lib/awful/rules.lua (hunk 13 above)
Comment (Elv13) [inline lib/awful/rules.lua]: Nothing new in the module. I still have my `managing-clients.md.lua` doc pages, but it's incomplete. It started from the same code that ended up being `appearance.md.lua`. It parses the code (client.c and client.lua) to build a property database then add the "custom" ones.

But for now I added a new commit with a crappy header doc. It also adds all the other undocumented properties.

**Comment by Elv13** [comment]
This is ready for another review and/or merge+open_issue for the remaining "problem" (that isn't really a regression)

**Comment by Elv13** [comment]
/me I think I will hear about these changes for some time... Hopefully they get us closer to something that works.

File: objects/client.c [hunk 15]
```diff
@@ -805,9 +805,17 @@
  * @function set_newindex_miss_handler
  */
 
+typedef enum {
+    CLIENT_MAXIMIZED_NONE = 0 << 0,
+    CLIENT_MAXIMIZED_V    = 1 << 0,
+    CLIENT_MAXIMIZED_H    = 1 << 1,
+    CLIENT_MAXIMIZED_BOTH = 1 << 2, /* V|H == BOTH, but ~(V|H) != ~(BOTH)... */
```
Comment (blueyed) [inline objects/client.c]: @Elv13 
The comment is not true (or I do not get it).. CLIENT_MAXIMIZED_BOTH is 4, but V|H is 3 (V being 1 and H being 0)?!
So `V|H == BOTH` is not true.

Comment (Elv13) [inline objects/client.c]: I was referring to the concepts. When you maximize both H and V, then you are *in effect* maximized. But when you undo the operation, you are left in a potentially inconsistent state.

I agree the comment is unhelpful and misleading. I should copy/paste the [discussion comment](https://github.com/awesomeWM/awesome/pull/1453#discussion_r97938954) into the code and remove this comment.
